src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -1305,23 +1317,25 @@ void CodeGen::genSSE41Intrinsic(GenTreeHWIntrinsic* node)
             {
                 tmpTargetReg = node->ExtractTempReg();
             }
-            auto emitSwCase = [&](unsigned i) {
+
+            auto emitSwCase = [&](int8_t i) {
                 if (baseType == TYP_FLOAT)
                 {
                     // extract instructions return to GP-registers, so it needs int size as the emitsize
-                    emit->emitIns_SIMD_R_R_I(ins, emitTypeSize(TYP_INT), tmpTargetReg, op1Reg, (int)i);
+                    emit->emitIns_SIMD_R_R_I(ins, emitTypeSize(TYP_INT), tmpTargetReg, op1Reg, i);
                     emit->emitIns_R_R(INS_mov_i2xmm, EA_4BYTE, targetReg, tmpTargetReg);
                 }
                 else
                 {
-                    emit->emitIns_SIMD_R_R_I(ins, emitTypeSize(TYP_INT), targetReg, op1Reg, (int)i);
+                    emit->emitIns_SIMD_R_R_I(ins, emitTypeSize(TYP_INT), targetReg, op1Reg, i);
                 }
             };
 
             if (op2->IsCnsIntOrI())
             {
                 ssize_t ival = op2->AsIntCon()->IconValue();
-                emitSwCase((unsigned)ival);
+                assert((ival >= 0) && (ival <= 255));
+                emitSwCase((int8_t)ival);
```

Comment:
It looks odd that you are effectively checking unsigned values, and then casting it to a signed type. Wouldn't it make more sense for the parameter to be declared unsigned? Is that because the emitter expects a signed type? Or is there another reason?

Reply:
Right. We take an unsigned type from our API, but the JIT expects it to be signed (the `fits in byte` code everywhere is basically `(signed char)val == val`).

I actually wonder if this causes us to generate code that is "too big" in some cases (several instructions don't care if the value is signed or unsigned and instead just set the right flags for both). But I haven't dug in any further to follow up on this.

Reply:
Ex: https://github.com/dotnet/coreclr/blob/master/src/jit/emitxarch.cpp#L10707

Just searching for `inByte` comes up with more examples, however.

It might also be worth noting that `GenTreeIntCon` sees to be the opposite. It checks for `0-255` to determine if we need to be a "small" or "large" constant (slightly different example, however, since this is mostly just compressing bits to save space)

Reply:
Right - it seems like this is an area where we could use some more clarity and checking ...